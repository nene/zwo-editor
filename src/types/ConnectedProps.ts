type AnyFn = (...args: any) => any;

type ReturnVoid<F extends AnyFn> = (...args: Parameters<F>) => void;

type FunctionMap = { [key: string]: AnyFn };

type VoidFunctionMap<T extends FunctionMap> = {
  [P in keyof T]: ReturnVoid<T[P]>;
};

// Derives Props type from mapStateToProps and mapDispatchToProps
export type ConnectedProps<
  MapStateToProps extends AnyFn,
  MapDispatchToProps extends FunctionMap = {}
> = ReturnType<MapStateToProps> & VoidFunctionMap<MapDispatchToProps>;
