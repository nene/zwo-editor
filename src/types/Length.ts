export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;

// Alternative approach
//
// This way we could avoid storing class instances to Redux,
// which eliminates problems with serialization & deserialization.
/*
export type Distance = { meters: number };
export type Duration = { seconds: number };
export type Length = Distance | Duration;

export const Distance = (meters: number) => ({ meters });
export const Duration = (seconds: number) => ({ seconds });

export const isDistance = (obj: Length): obj is Distance => 'meters' in obj;
export const isDuration = (obj: Length): obj is Duration => 'seconds' in obj;
*/
