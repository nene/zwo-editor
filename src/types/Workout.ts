import { SportType } from "./SportType";
import { Interval } from "./Interval";
import { LengthType } from "./LengthType";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  lengthType: LengthType;
  tags: string[];
  intervals: Array<Interval>;
}

export function createEmptyWorkout(
  sportType: SportType,
  lengthType: LengthType
): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: sportType,
    lengthType: lengthType,
    tags: [],
    intervals: [],
  };
}
