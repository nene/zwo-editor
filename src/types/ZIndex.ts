export enum ZIndex {
  fader = 1,
  selected = 5,
  instruction = 11,
  tooltip = 100,
  popup = 1000,
}
