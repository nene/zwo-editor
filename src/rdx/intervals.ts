import { createSlice, createSelector, PayloadAction } from '@reduxjs/toolkit';
import { REHYDRATE } from 'redux-persist';
import { Interval } from '../types/Interval';
import { RootState } from './store';
import { convertLengths } from '../storage/storage';

interface IntervalsState {
  intervals: Interval[];
}

const initialState: IntervalsState = { intervals: [] };

const intervalsSlice = createSlice({
  name: 'intervals',
  initialState,
  reducers: {
    setIntervals: (state, action: PayloadAction<Interval[]>) => ({ intervals: action.payload }),
  },
  extraReducers: {
    [REHYDRATE]: (state, action: PayloadAction<{intervals: IntervalsState}>) => {
      if (!action.payload.intervals) {
        return state;
      }
      return {
        intervals: action.payload.intervals.intervals.map(convertLengths) as Interval[],
      };
    },
  },
});

export const reducer = intervalsSlice.reducer;
export const { setIntervals } = intervalsSlice.actions;

const selectBase = (state: RootState) => state.intervals;
export const selectIntervals = createSelector(selectBase, (s) => s.intervals);
