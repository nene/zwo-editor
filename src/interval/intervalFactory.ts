import { v4 as uuidv4 } from "uuid";
import { PaceType } from "../types/PaceType";
import {
  FreeInterval,
  Interval,
  RampInterval,
  RepetitionInterval,
  SteadyInterval,
} from "../types/Interval";
import { Distance, Duration, Length } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";

const defaultDuration = Duration(300);
const defaultDistance = Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: "steady",
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: "ramp",
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: "free",
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: undefined,
      instructions: [],
      ...interval,
    };
  },

  repetition(
    interval: Partial<RepetitionInterval>,
    mode: WorkoutMode
  ): RepetitionInterval {
    return {
      type: "repetition",
      id: uuidv4(),
      repeat: 3,
      onLength: mode.lengthType === "time" ? Duration(30) : Distance(200),
      offLength: mode.lengthType === "time" ? Duration(120) : Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      onCadence: undefined,
      offCadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
