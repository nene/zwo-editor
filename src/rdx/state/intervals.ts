import {
  createAsyncThunk,
  createSelector,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import { Interval } from "../../types/Interval";
import { RootState } from "../store";
import { clearWorkout, loadWorkout } from "./workout";
import {
  updateIntervalIntensity,
  updateIntervalDuration,
  moveInterval,
} from "../../interval/intervalUtils";
import { replaceById } from "../../utils/array";
import { clearSelection, selectSelectedId } from "./selectedId";
import intervalFactory from "../../interval/intervalFactory";
import { Duration } from "../../types/Length";
import { selectMode } from "./mode";
import { PaceType } from "../../types/PaceType";
import { not, propEq } from "ramda";

const initialState: Interval[] = [];

const slice = createSlice({
  name: "intervals",
  initialState,
  reducers: {
    setIntervals: (state, action: PayloadAction<Interval[]>) => action.payload,
    addInterval: (state, action: PayloadAction<Interval>) => [
      ...state,
      action.payload,
    ],
    updateInterval: (intervals, { payload }: PayloadAction<Interval>) =>
      replaceById(payload, intervals),
    removeInterval: (
      intervals,
      { payload: id }: PayloadAction<string | undefined>
    ) => intervals.filter((item) => item.id !== id),
    removeInstruction: (
      intervals,
      { payload }: PayloadAction<{ intervalId: string; instructionId: string }>
    ) => {
      const interval = intervals.find(propEq("id", payload.intervalId));
      if (interval) {
        return replaceById(
          {
            ...interval,
            instructions: interval.instructions.filter(
              (i) => i.id !== payload.instructionId
            ),
          },
          intervals
        );
      }
      return intervals;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(clearWorkout, () => [])
      .addCase(loadWorkout, (state, { payload }) => payload.intervals);
  },
});

export const reducer = slice.reducer;
export const {
  setIntervals,
  addInterval,
  updateInterval,
  removeInterval,
  removeInstruction,
} = slice.actions;

export const removeSelectedInterval = createAsyncThunk(
  "intervals/removeSelectedInterval",
  (_: void, { getState, dispatch }) => {
    const selectedId = selectSelectedId(getState() as RootState);
    dispatch(removeInterval(selectedId));
    dispatch(clearSelection());
  }
);

export const duplicateSelectedInterval = createAsyncThunk(
  "intervals/duplicateSelectedInterval",
  (_: void, { getState, dispatch }) => {
    const interval = selectSelectedInterval(getState() as RootState);
    if (interval) {
      dispatch(addInterval(intervalFactory.clone(interval)));
    }
    dispatch(clearSelection());
  }
);

export const adjustSelectedIntervalIntensity = createAsyncThunk(
  "intervals/adjustSelectedIntervalIntensity",
  (amount: number, { getState, dispatch }) => {
    const selectedId = selectSelectedId(getState() as RootState);
    if (selectedId) {
      const intervals = selectIntervals(getState() as RootState);
      dispatch(
        setIntervals(updateIntervalIntensity(selectedId, amount, intervals))
      );
    }
  }
);

export const adjustSelectedIntervalDuration = createAsyncThunk(
  "intervals/adjustSelectedIntervalDuration",
  (amount: Duration, { getState, dispatch }) => {
    const selectedId = selectSelectedId(getState() as RootState);
    if (selectedId) {
      const intervals = selectIntervals(getState() as RootState);
      const mode = selectMode(getState() as RootState);
      dispatch(
        setIntervals(
          updateIntervalDuration(selectedId, amount, intervals, mode)
        )
      );
    }
  }
);

export const moveSelectedInterval = createAsyncThunk(
  "intervals/moveSelectedInterval",
  (direction: -1 | 1, { getState, dispatch }) => {
    const selectedId = selectSelectedId(getState() as RootState);
    if (selectedId) {
      const intervals = selectIntervals(getState() as RootState);
      dispatch(setIntervals(moveInterval(selectedId, direction, intervals)));
    }
  }
);

export const setSelectedIntervalPace = createAsyncThunk(
  "intervals/setSelectedIntervalPace",
  (pace: PaceType, { getState, dispatch }) => {
    const interval = selectSelectedInterval(getState() as RootState);
    if (interval && interval.type === "steady") {
      // TODO: Only steady?
      dispatch(updateInterval({ ...interval, pace }));
    }
  }
);

export const selectIntervals = (state: RootState) => state.intervals;

export const selectSelectedInterval = createSelector(
  selectSelectedId,
  selectIntervals,
  (selectedId, intervals) => {
    return intervals.find((interval) => interval.id === selectedId);
  }
);

export const selectSelectedIntervalPace = createSelector(
  selectSelectedInterval,
  (interval) => {
    return interval && interval.type !== "free" ? interval.pace : undefined;
  }
);
