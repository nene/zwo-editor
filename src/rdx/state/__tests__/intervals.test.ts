import intervalFactory from "../../../interval/intervalFactory";
import BikeMode from "../../../modes/BikeMode";
import { Duration } from "../../../types/Length";
import { reducer, removeInterval } from "../intervals";

const defaultMode = () => new BikeMode(200, 75);

describe("intervals reducer", () => {
  describe("removeInterval()", () => {
    it("removes interval with given ID", () => {
      const state = [
        intervalFactory.steady(
          {
            id: "#1",
            length: Duration(60),
            intensity: 0.5,
            instructions: [],
          },
          defaultMode()
        ),
        intervalFactory.steady(
          {
            id: "#2",
            length: Duration(60),
            intensity: 0.5,
            instructions: [],
          },
          defaultMode()
        ),
      ];
      expect(reducer(state, removeInterval("#2"))).toMatchSnapshot();
    });
  });
});
