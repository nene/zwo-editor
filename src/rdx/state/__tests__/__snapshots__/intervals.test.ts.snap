// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intervals reducer removeInterval() removes interval with given ID 1`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;
